core/src/main/scala/org/broadinstitute/dsde/rawls/fastpass/FastPassService.scala

```diff
@@ -0,0 +1,444 @@
+            workbenchEmail,
+            memberType,
+            organizationRoles,
+            userProject = Some(googleProject)
+          )
+        case IamResourceTypes.Project =>
+          googleIamDao.removeRoles(
+            googleProject,
+            workbenchEmail,
+            memberType,
+            organizationRoles
+          )
+      }
+      _ <- openTelemetry
+        .incrementCounter(s"fastpass-iam-revoked-${memberType.value}", tags = openTelemetryTags)
+        .unsafeToFuture()
+    } yield ()
+  }
+
+  private def setupProjectRoles(workspace: Workspace,
+                                samResourceRoles: Set[SamResourceRole],
+                                userAndPet: UserAndPetEmails,
+                                expirationDate: DateTime
+  ): ReadWriteAction[Unit] = {
+    val projectIamRoles = samResourceRoles.flatMap(samWorkspaceRoleToGoogleProjectIamRoles)
+    val condition = conditionFromExpirationDate(expirationDate)
+
+    for {
+      _ <- DBIO.from(addUserAndPetToProjectIamRoles(workspace.googleProjectId, projectIamRoles, userAndPet, condition))
+      _ <- writeGrantsToDb(
+        workspace.workspaceId,
+        userAndPet,
+        gcpResourceType = IamResourceTypes.Project,
+        workspace.googleProjectId.value,
+        projectIamRoles,
+        expirationDate
+      )
+    } yield ()
+  }
+
+  private def setupBucketRoles(workspace: Workspace,
+                               samResourceRoles: Set[SamResourceRole],
+                               userAndPet: UserAndPetEmails,
+                               expirationDate: DateTime
+  ): ReadWriteAction[Unit] = {
+    val bucketIamRoles = samResourceRoles.flatMap(samWorkspaceRolesToGoogleBucketIamRoles)
+    val condition = conditionFromExpirationDate(expirationDate)
+
+    for {
+      _ <- DBIO.from(
+        addUserAndPetToBucketIamRole(GcsBucketName(workspace.bucketName),
+                                     bucketIamRoles,
+                                     userAndPet,
+                                     condition,
+                                     workspace.googleProjectId
+        )
+      )
+      _ <- writeGrantsToDb(
+        workspace.workspaceId,
+        userAndPet,
+        gcpResourceType = IamResourceTypes.Bucket,
+        workspace.bucketName,
+        bucketIamRoles,
+        expirationDate
+      )
+    } yield ()
+  }
+
+  private def writeGrantsToDb(workspaceId: String,
+                              userAndPet: UserAndPetEmails,
+                              gcpResourceType: IamResourceType,
+                              resourceName: String,
+                              organizationRoles: Set[String],
+                              expiration: DateTime
+  ): ReadWriteAction[Unit] = {
+    val rolesToWrite =
+      Seq((userAndPet.userEmail, IamMemberTypes.User), (userAndPet.petEmail, IamMemberTypes.ServiceAccount)).flatMap(
+        tuple => organizationRoles.map(r => (tuple._1, tuple._2, r))
+      )
```

Comment:
nit - This took me a little while to grok.  This statement is doing quite a few things.  Suggest adding a type declaration onto the `val`.  Also, in the `organizationRoles.map` the fact that it was just called `r` meant that my brain kind of missed it.  Maybe it's just me, but I thought this thing was making a `Tuple2` because of the `r` and the lack of a type declaration.  I don't know, maybe this would read better if you `flatMap` over the `organizationRoles` instead of the `Seq[Tuple2]`?  I don't know.